File: server/src/main/java/com/nexoffer/controller/BookmarkController.java
package com.nexoffer.controller;

import com.nexoffer.model.Bookmark;
import com.nexoffer.model.User;
import com.nexoffer.repository.BookmarkRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/bookmarks")
public class BookmarkController {

    private final BookmarkRepository bookmarkRepository;

    public BookmarkController(BookmarkRepository bookmarkRepository) {
        this.bookmarkRepository = bookmarkRepository;
    }

    @GetMapping
    public ResponseEntity<?> getBookmarks(@AuthenticationPrincipal User user,
                                          @RequestParam(required = false) String category,
                                          @RequestParam(required = false) String search) {
        List<Bookmark> bookmarks;
        if (category != null && !category.equalsIgnoreCase("All")) {
            bookmarks = bookmarkRepository.findByUserIdAndCategoryOrderByCreatedAtDesc(user.getId(), category);
        } else {
            bookmarks = bookmarkRepository.findByUserIdOrderByCreatedAtDesc(user.getId());
        }

        if (search != null && !search.isBlank()) {
            String lower = search.toLowerCase();
            bookmarks = bookmarks.stream()
                    .filter(b -> (b.getQuestion() != null && b.getQuestion().toLowerCase().contains(lower)) ||
                                 (b.getTopic() != null && b.getTopic().toLowerCase().contains(lower)) ||
                                 (b.getNotes() != null && b.getNotes().toLowerCase().contains(lower)))
                    .toList();
        }

        return ResponseEntity.ok(Map.of(
                "success", true,
                "count", bookmarks.size(),
                "bookmarks", bookmarks
        ));
    }

    @PostMapping
    public ResponseEntity<?> createBookmark(@AuthenticationPrincipal User user, @RequestBody Map<String, Object> body) {
        String question = body.get("question") != null ? body.get("question").toString() : null;
        if (question == null || question.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Question content is required."));
        }

        Optional<Bookmark> existing = bookmarkRepository.findByUserIdAndQuestion(user.getId(), question.trim());
        if (existing.isPresent()) {
            return ResponseEntity.ok(Map.of(
                    "success", true,
                    "message", "Question is already in your bookmarks.",
                    "bookmark", existing.get()
            ));
        }

        String category = body.get("category") != null ? body.get("category").toString() : "General";
        String difficulty = body.get("difficulty") != null ? body.get("difficulty").toString() : "Medium";
        String topic = body.get("topic") != null ? body.get("topic").toString() : "";
        String suggestedAnswer = body.get("suggestedAnswer") != null ? body.get("suggestedAnswer").toString() : "";
        String notes = body.get("notes") != null ? body.get("notes").toString() : "";

        Bookmark bookmark = new Bookmark(
                user.getId(),
                question.trim(),
                category,
                difficulty,
                topic,
                suggestedAnswer,
                notes
        );

        Bookmark saved = bookmarkRepository.save(bookmark);

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                "success", true,
                "message", "Question saved to bookmarks!",
                "bookmark", saved,
                "data", saved
        ));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateBookmark(@AuthenticationPrincipal User user,
                                            @PathVariable Long id,
                                            @RequestBody Map<String, Object> body) {
        Optional<Bookmark> bookmarkOpt = bookmarkRepository.findByIdAndUserId(id, user.getId());
        if (bookmarkOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("success", false, "message", "Bookmark not found."));
        }

        Bookmark bookmark = bookmarkOpt.get();
        if (body.containsKey("notes")) {
            bookmark.setNotes(body.get("notes") != null ? body.get("notes").toString() : "");
        }
        Bookmark saved = bookmarkRepository.save(bookmark);

        return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Bookmark updated successfully!",
                "bookmark", saved
        ));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteBookmark(@AuthenticationPrincipal User user, @PathVariable Long id) {
        Optional<Bookmark> bookmarkOpt = bookmarkRepository.findByIdAndUserId(id, user.getId());
        if (bookmarkOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("success", false, "message", "Bookmark not found."));
        }

        bookmarkRepository.delete(bookmarkOpt.get());

        return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Bookmark removed successfully."
        ));
    }
}

File: server/controllers/bookmarkController.js
const { BookmarkRepository } = require('../models/dbStore');

// @desc    Get all bookmarked questions for current user
// @route   GET /api/bookmarks
const getBookmarks = async (req, res, next) => {
  try {
    const { category, search } = req.query;
    const query = { userId: req.user._id };

    if (category && category !== 'All') {
      query.category = category;
    }

    if (search) {
      query.question = { $regex: search, $options: 'i' };
    }

    const cursor = await BookmarkRepository.find(query);
    const bookmarks = cursor.sort ? await cursor.sort({ createdAt: -1 }) : cursor;

    res.status(200).json({
      success: true,
      count: bookmarks.length,
      bookmarks,
    });
  } catch (error) {
    next(error);
  }
};

// @desc    Add a question to bookmarks
// @route   POST /api/bookmarks
const createBookmark = async (req, res, next) => {
  try {
    const { question, category = 'General', difficulty = 'Medium', topic = '', suggestedAnswer = '', notes = '' } = req.body;

    if (!question || question.trim().length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Question content is required.',
      });
    }

    const existing = await BookmarkRepository.findOne({
      userId: req.user._id,
      question: question.trim(),
    });

    if (existing) {
      return res.status(200).json({
        success: true,
        message: 'Question is already in your bookmarks.',
        bookmark: existing,
      });
    }

    const bookmark = await BookmarkRepository.create({
      userId: req.user._id,
      question: question.trim(),
      category,
      difficulty,
      topic,
      suggestedAnswer,
      notes,
    });

    res.status(201).json({
      success: true,
      message: 'Question saved to bookmarks!',
      bookmark,
    });
  } catch (error) {
    next(error);
  }
};

// @desc    Update a bookmark (notes)
// @route   PUT /api/bookmarks/:id
const updateBookmark = async (req, res, next) => {
  try {
    const { notes } = req.body;
    const bookmark = await BookmarkRepository.findOne({
      _id: req.params.id,
      userId: req.user._id,
    });

    if (!bookmark) {
      return res.status(404).json({
        success: false,
        message: 'Bookmark not found.',
      });
    }

    if (notes !== undefined) bookmark.notes = notes;
    await bookmark.save();

    res.status(200).json({
      success: true,
      message: 'Bookmark updated successfully!',
      bookmark,
    });
  } catch (error) {
    next(error);
  }
};

// @desc    Delete a bookmark
// @route   DELETE /api/bookmarks/:id
const deleteBookmark = async (req, res, next) => {
  try {
    const bookmark = await BookmarkRepository.findOneAndDelete({
      _id: req.params.id,
      userId: req.user._id,
    });

    if (!bookmark) {
      return res.status(404).json({
        success: false,
        message: 'Bookmark not found.',
      });
    }

    res.status(200).json({
      success: true,
      message: 'Bookmark removed successfully.',
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getBookmarks,
  createBookmark,
  updateBookmark,
  deleteBookmark,
};
